XMonad/Layout/Decoration.hs

```diff
@@ -400,7 +400,11 @@ updateDecos s t f = mapM_ $ updateDeco s t f
 -- structure and the needed 'Rectangle's
 updateDeco :: Shrinker s => s -> Theme -> XMonadFont -> (OrigWin,DecoWin) -> X ()
 updateDeco sh t fs ((w,_),(Just dw,Just (Rectangle _ _ wh ht))) = do
-  nw  <- getName w
+  -- xmonad-contrib #809
+  -- qutebrowser will happily shovel a 389K multiline string into @_NET_WM_NAME@
+  -- and the 'defaultShrinker' (a) doesn't handle multiline strings well (b) is
+  -- quadratic due to using 'init'
```

Comment:
I didn't completely follow all of this, but is there any reason why `defaultShrinker` *should* be quadratic or is that a bug we ought to fix as well?

Reply:
It shrinks by a character at a time until the rendered version fits inside the `Decoration`. Unfortunately, you can't run that one backwards and determine the proper length of the `String` from what will fit inside the `Decoration` directly, because even if you can ensure that a monospaced font was used (we can't; that's only the default) there are Unicode glyphs that take up the space of two monospaced characters.

Reply:
But can't you do like a binary search over the length of the string? I don't know how often you only have to chop of one or two characters; I feel like when you have to chop, you have to chop quite a lot

Reply:
Probably. It would also partially avoid problems like splitting ligatures (flags, etc.). But I feel like that's beyond me.